crates/core/src/routing_info/mod.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+use std::pin::pin;
+use std::sync::Arc;
+
+use anyhow::Context;
+use arc_swap::ArcSwap;
+use tokio::sync::{mpsc, oneshot};
+use tokio::time::MissedTickBehavior;
+use tracing::{debug, info, trace};
+use xxhash_rust::xxh3::Xxh3Builder;
+
+use restate_types::cluster_controller::SchedulingPlan;
+use restate_types::config::Configuration;
+use restate_types::identifiers::PartitionId;
+use restate_types::metadata_store::keys::SCHEDULING_PLAN_KEY;
+use restate_types::NodeId;
+
+use crate::metadata_store::MetadataStoreClient;
+use crate::{
+    cancellation_watcher, task_center, ShutdownError, TargetVersion, TaskCenter, TaskHandle,
+    TaskId, TaskKind,
+};
+
+pub type CommandSender = mpsc::Sender<Command>;
+pub type CommandReceiver = mpsc::Receiver<Command>;
+
+pub enum Command {
+    /// Request an on-demand refresh of routing information from the authoritative source.
+    SyncRoutingInformation(TargetVersion),
+}
+
+/// Holds a view of the known partition-to-node mappings. Use it to discover which node(s) to route
+/// requests to for a given partition. Compared to the partition table, this view is more dynamic as
+/// it changes based on cluster nodes' operational status. This handle can be cheaply cloned.
+#[derive(Clone)]
+pub struct PartitionRouting {
```

Comment:
I'm not in love with this name but hate everything else I've come up with more. Any suggestions? `PartitionNodeRoutes` is a more verbose version I also like.